### PR TITLE
Added support for passing S3 endpoint as an argument to dataset construction API to access dataset stored in AWS S3 compatible storage systems

### PR DESCRIPTION
**Pull request by bharatendra**
## Description
Provide a way to pass endpoint as an optional argument to the dataset manipulation API classes S3MapDataset, S3IterableDataset, and S3Checkpoint. To support this new optional argument, modified Rust client code MountpointS3Client. By default the argument is set to empty string. In the  MountpointS3Client, the endpoint is set in EndpointConfig if it is not empty. With this change the pytorch s3 connector can work with AWS S3 compatible storage systems.

## Additional context
Current pytorch s3 connector can only work with AWS S3 service as it by default connects to AWS S3 endpoint. Only the region parameter can be passed. In order to work with datasets stored in AWS S3 compatible storage systems, endpoint needs to be changed. 

## Testing
Tested with example code of accessing datasets using S3MapDataset, S3IterableDataset and saving checkpoints using S3Checkpoint. Tested with empty and non-empty endpoint argument.
--------
By submitting this pull request, I confirm that my contribution is made under the terms of BSD 3-Clause License and I agree to the terms of the [LICENSE](https://github.com/awslabs/s3-connector-for-pytorch/blob/main/LICENSE).

### CODE REVIEW

**Comment by ASchneidman** [comment]
I'd like to bump this PR, would be very useful for our codebase as well.

File: s3torchconnector/src/s3torchconnector/s3map_dataset.py [hunk 1]
```diff
@@ -26,19 +26,25 @@ class S3MapDataset(torch.utils.data.Dataset):
     def __init__(
         self,
         region: str,
+        endpoint: str,
```
Comment (muddyfish) [inline s3torchconnector/src/s3torchconnector/s3map_dataset.py:29]: In `S3IterableDataset`, this parameter has a default value (`""`). It should be consistent between these two classes, and be in the same order.

File: s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi [hunk 2]
```diff
@@ -16,6 +16,7 @@ class MountpointS3Client:
     def __init__(
         self,
         region: str,
+        endpoint: str,
```
Comment (muddyfish) [inline s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi:19]: This has a default (`""`)

File: s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi [hunk 3]
```diff
@@ -35,6 +36,7 @@ class MockMountpointS3Client:
...
```
Comment (muddyfish) [inline s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi:39]: `MockMountpointS3Client` was left without an endpoint parameter. It probably should have one.

File: s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi [hunk 4]
```diff
@@ -16,6 +16,7 @@ class MountpointS3Client:
...
         user_agent_prefix: str = "",
```
Comment (muddyfish) [inline s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi:20]: A new `endpoint` read only attribute was added to this class, it should be added to the list of attributes at the top

**Comment by muddyfish** [comment]
Hi @bharatendra, thanks for the contribution. Would you be able to add some basic unit tests to this as well (creating an S3MountpointClient with an endpoint gives you the same endpoint when you get the attribute, same for `S3Checkpoint`, `MapDataset`, `IterableDataset` keeping the endpoint attribute and passing it through)

**Review by muddyfish — CHANGES_REQUESTED** [review]

File: s3torchconnector/src/s3torchconnector/s3checkpoint.py [hunk 5]
```diff
@@ -16,8 +16,9 @@ class S3Checkpoint:
     torch.load, and torch.save.
     """
 
-    def __init__(self, region: str):
+    def __init__(self, region: str, endpoint: str = ""):
         self.region = region
+        self.endpoint = endpoint
         self._client = S3Client(region)
```
[the remark this answers is not in the document]
Comment (bharatendra) [reply s3torchconnector/src/s3torchconnector/s3checkpoint.py]: Thank you for the review. I made the change to pass endpoint to S3Client creation

File: s3torchconnector/src/s3torchconnector/s3map_dataset.py (hunk 1 above)
Comment (bharatendra) [reply s3torchconnector/src/s3torchconnector/s3map_dataset.py]: I made the S3MapDataset class consistent with S3IterableDataset

File: s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi (hunk 3 above)
Comment (bharatendra) [reply s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi]: I made the endpoint assigned with empty string by default

File: s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi (hunk 4 above)
Comment (bharatendra) [reply s3torchconnectorclient/python/src/s3torchconnectorclient/_mountpoint_s3_client.pyi:21]: Added endpoint to class member variable

**Comment by muddyfish** [comment]
Thanks for the changes, once the linting issue is fixed, we're happy to merge this in.

File: s3torchconnectorclient/rust/src/mountpoint_s3_client.rs [hunk 6]
```diff
@@ -42,26 +44,33 @@ pub struct MountpointS3Client {
...
-    #[pyo3(signature = (region, user_agent_prefix="".to_string(), throughput_target_gbps=10.0, part_size=8*1024*1024, profile=None, no_sign_request=false))]
+    #[pyo3(signature = (region, user_agent_prefix="".to_string(), throughput_target_gbps=10.0, part_size=8*1024*1024, profile=None, no_sign_request=false, endpoint="".to_string()))]
     pub fn new_s3_client(
         region: String,
         user_agent_prefix: String,
         throughput_target_gbps: f64,
         part_size: usize,
         profile: Option<String>,
         no_sign_request: bool,
+        endpoint: String,
```
Comment (muddyfish) [inline s3torchconnectorclient/rust/src/mountpoint_s3_client.rs:64]: Can we change this to be an `Option<String>` to be consistent with `profile`?

File: s3torchconnectorclient/rust/src/mountpoint_s3_client.rs [hunk 7]
```diff
@@ -42,26 +44,33 @@ pub struct MountpointS3Client {
...
+    endpoint: String,
 
     owner_pid: Pid,
 }
 
 #[pymethods]
 impl MountpointS3Client {
     #[new]
-    #[pyo3(signature = (region, user_agent_prefix="".to_string(), throughput_target_gbps=10.0, part_size=8*1024*1024, profile=None, no_sign_request=false))]
+    #[pyo3(signature = (region, user_agent_prefix="".to_string(), throughput_target_gbps=10.0, part_size=8*1024*1024, profile=None, no_sign_request=false, endpoint="".to_string()))]
```
Comment (muddyfish) [inline s3torchconnectorclient/rust/src/mountpoint_s3_client.rs:56]: Can we also change this to be default `None`

File: s3torchconnector/src/s3torchconnector/_s3client/_s3client.py [hunk 8]
```diff
@@ -28,8 +28,9 @@ def _identity(obj: Any) -> Any:
 
 
 class S3Client:
-    def __init__(self, region: str):
+    def __init__(self, region: str, endpoint: str = ""):
```
Comment (muddyfish) [inline s3torchconnector/src/s3torchconnector/_s3client/_s3client.py:31]: Can we change all instances of endpoint default to be `None` rather than empty string?

File: s3torchconnector/src/s3torchconnector/s3checkpoint.py [hunk 9]
```diff
@@ -16,9 +16,10 @@ class S3Checkpoint:
     torch.load, and torch.save.
     """
 
-    def __init__(self, region: str):
+    def __init__(self, region: str, endpoint: str = ""):
```
Comment (muddyfish) [inline s3torchconnector/src/s3torchconnector/s3checkpoint.py:19]: Same here for default `None` and other occurances

File: s3torchconnector/src/s3torchconnector/s3iterable_dataset.py [hunk 10]
```diff
@@ -26,30 +26,38 @@ def __init__(
...
+    def endpoint(self):
+        return self._endpoint
+
     @classmethod
     def from_objects(
         cls,
         object_uris: Union[str, Iterable[str]],
         *,
         region: str,
+        endpoint: str = "",
```
Comment (muddyfish) [inline s3torchconnector/src/s3torchconnector/s3iterable_dataset.py:52]: Same here for default `None`

File: s3torchconnector/src/s3torchconnector/s3iterable_dataset.py [hunk 11]
```diff
@@ -68,13 +79,15 @@ def from_prefix(
         s3_uri: str,
         *,
         region: str,
+        endpoint: str = "",
```
Comment (muddyfish) [inline s3torchconnector/src/s3torchconnector/s3iterable_dataset.py:82]: Same here for default `None`

File: s3torchconnectorclient/rust/src/mock_client.rs [hunk 12]
```diff
@@ -57,6 +57,7 @@ impl PyMockClient {
             None,
             false,
             self.mock_client.clone(),
+            "".to_string(),
```
Comment (muddyfish) [inline s3torchconnectorclient/rust/src/mock_client.rs:60]: Here can we change to `None` as well?

File: s3torchconnectorclient/rust/src/mountpoint_s3_client.rs [hunk 13]
```diff
@@ -42,26 +44,33 @@ pub struct MountpointS3Client {
...
+        endpoint: String,
     ) -> PyResult<Self> {
         // TODO: Mountpoint has logic for guessing based on instance type. It may be worth having
         // similar logic if we want to exceed 10Gbps reading for larger instances
 
-        let endpoint_config = EndpointConfig::new(&region);
+        let endpoint_config = if endpoint.is_empty() {
+            EndpointConfig::new(&region)
+        } else {
+            EndpointConfig::new(&region).endpoint(Uri::new_from_str(&Allocator::default(), endpoint.clone()).unwrap())
```
Comment (muddyfish) [inline s3torchconnectorclient/rust/src/mountpoint_s3_client.rs:72]: Can we see if we need to make changes here if an invalid URI is passed? (And associated unit tests).

**Review by muddyfish — CHANGES_REQUESTED** [review]

File: s3torchconnectorclient/rust/src/mountpoint_s3_client.rs (hunk 6 above)
Comment (bharatendra) [reply s3torchconnectorclient/rust/src/mountpoint_s3_client.rs]: Changed the type of endpoint parameter to Option<String>

File: s3torchconnectorclient/rust/src/mountpoint_s3_client.rs (hunk 7 above)
Comment (bharatendra) [reply s3torchconnectorclient/rust/src/mountpoint_s3_client.rs]: Changed the default value of endpoint parameter to None instead of empty string

File: s3torchconnector/src/s3torchconnector/s3checkpoint.py (hunk 9 above)
Comment (bharatendra) [reply s3torchconnector/src/s3torchconnector/s3checkpoint.py]: Changed the default value of endpoint parameter to None instead of empty string

File: s3torchconnector/src/s3torchconnector/s3iterable_dataset.py (hunk 10 above)
Comment (bharatendra) [reply s3torchconnector/src/s3torchconnector/s3iterable_dataset.py]: Changed the default value of endpoint parameter to None instead of empty string

File: s3torchconnector/src/s3torchconnector/s3iterable_dataset.py (hunk 11 above)
Comment (bharatendra) [reply s3torchconnector/src/s3torchconnector/s3iterable_dataset.py]: Changed the default value of endpoint parameter to None instead of empty string

File: s3torchconnectorclient/rust/src/mountpoint_s3_client.rs (hunk 13 above)
Comment (bharatendra) [reply s3torchconnectorclient/rust/src/mountpoint_s3_client.rs]: Changed the default value of endpoint parameter to None instead of empty string. If an invalid URI is passed the underneath code in moutpoits3 will report error

**Comment by muddyfish** [comment]
Thanks for the changes to make the endpoint `Optional`. One nitpick though, could you please add some unit tests for when the endpoint is None/empty string/invalid URI/valid URI (I think you had the last already)

File: s3torchconnectorclient/python/tst/unit/test_mountpoint_s3_client.py [hunk 14]
```diff
@@ -25,6 +25,9 @@
 REGION = "us-east-1"
 MOCK_BUCKET = "mock-bucket"
 ENDPOINT = "https://s3.us-east-1.amazonaws.com"
+EMPTY_ENDPOINT = ""
+NONE_ENDPOINT = None
+INVALID_ENDPOINT = "https://us-east-1.amazonaws.com"
```
Comment (muddyfish) [inline s3torchconnectorclient/python/tst/unit/test_mountpoint_s3_client.py:30]: Thanks for adding these tests. Can we cover these scenarios with a parameterized test such as in https://github.com/awslabs/s3-connector-for-pytorch/blob/main/s3torchconnector/tst/unit/test_s3mapdataset.py#L30-L37 ?
Can we also add another one with a completely invalid URI (say `"[INVALID]"`) with associated `pytest.raises` block? (Example https://github.com/awslabs/s3-connector-for-pytorch/blob/main/s3torchconnectorclient/python/tst/unit/test_mountpoint_s3_client.py#L241-L243)

Comment (bharatendra) [reply s3torchconnectorclient/python/tst/unit/test_mountpoint_s3_client.py]: I made the changes. For invalid URI case, I need to use the put_object method on MountpointS3Client as the URI validation occurs when trying to connect.

File: s3torchconnectorclient/python/tst/unit/test_mountpoint_s3_client.py [hunk 15]
```diff
@@ -4,7 +4,7 @@
 import logging
 import pickle
 import pytest
-from typing import Set
+from typing import Set, Sequence
```
Comment (muddyfish) [inline s3torchconnectorclient/python/tst/unit/test_mountpoint_s3_client.py:7]: Nit: This import isn't needed

**Comment by muddyfish** [comment]
Thanks for the new unit tests. We're planning on adding a new environment variable `CI_CUSTOM_ENDPOINT_URL` and associated VPC endpoint for this as part of our CI.


Would it also be possible to add an integration test using this new variable which verifies we can read the contents of an object using this custom endpoint?

File: s3torchconnectorclient/python/tst/unit/test_mountpoint_s3_client.py (hunk 15 above)
Comment (bharatendra) [reply s3torchconnectorclient/python/tst/unit/test_mountpoint_s3_client.py]: I removed the extra module import

**Comment by bharatendra** [comment]
> Thanks for the new unit tests. We're planning on adding a new environment variable `CI_CUSTOM_ENDPOINT_URL` and associated VPC endpoint for this as part of our CI.
> 
> Would it also be possible to add an integration test using this new variable which verifies we can read the contents of an object using this custom endpoint?

I added an integration test by setting endpoint to CI_CUSTOM_ENDPOINT_URL. But I am unable to test it as I am getting access denied error. Could you please look in to it

**Comment by dnanuti** [comment]
Hi @bharatendra! Thank you for your patience and sorry for the delay. I've pulled in your changes and tested them with the standard S3 endpoint for now, while we sort out the infrastructure on our end:
```
def test_get_object_with_endpoint(sample_directory):
    client = MountpointS3Client(
        sample_directory.region, TEST_USER_AGENT_PREFIX, endpoint="http://s3.amazonaws.com"
    )
    stream = client.get_object(
        sample_directory.bucket, f"{sample_directory.prefix}hello_world.txt"
    )

    full_data = b"".join(stream)
    assert full_data == HELLO_WORLD_DATA
```
The test has passed:
```
Launching pytest with arguments test_mountpoint_s3_integration.py::test_get_object_with_endpoint --no-header --no-summary -q in /Volumes/workplace/external-prs/s3-connector-for-pytorch-endpoint/s3torchconnectorclient/python/tst/integration

============================= test session starts ==============================
collecting ... collected 1 item

test_mountpoint_s3_integration.py::test_get_object_with_endpoint PASSED  [100%]

============================== 1 passed in 4.06s ===============================
```

I set the CI_CUSTOM_ENDPOINT_URL = http://s3.amazonaws.com as a variable for our actions as well, but we need updates on the Github workflow file from the main branch as well to take this into account.
Let's do the following: 
- Please resolve the merge conflicts
- Use the static value `endpoint="http://s3.amazonaws.com"` for these tests
And I'll take it on our end to update the infrastructure and CI accordingly. 
Is this ok with you?

**Comment by bharatendra** [comment]
> Hi @bharatendra! Thank you for your patience and sorry for the delay. I've pulled in your changes and tested them with the standard S3 endpoint for now, while we sort out the infrastructure on our end:
> 
> ```
> def test_get_object_with_endpoint(sample_directory):
>     client = MountpointS3Client(
>         sample_directory.region, TEST_USER_AGENT_PREFIX, endpoint="http://s3.amazonaws.com"
>     )
>     stream = client.get_object(
>         sample_directory.bucket, f"{sample_directory.prefix}hello_world.txt"
>     )
> 
>     full_data = b"".join(stream)
>     assert full_data == HELLO_WORLD_DATA
> ```
> 
> The test has passed:
> 
> ```
> Launching pytest with arguments test_mountpoint_s3_integration.py::test_get_object_with_endpoint --no-header --no-summary -q in /Volumes/workplace/external-prs/s3-connector-for-pytorch-endpoint/s3torchconnectorclient/python/tst/integration
> 
> ============================= test session starts ==============================
> collecting ... collected 1 item
> 
> test_mountpoint_s3_integration.py::test_get_object_with_endpoint PASSED  [100%]
> 
> ============================== 1 passed in 4.06s ===============================
> ```
> 
> I set the CI_CUSTOM_ENDPOINT_URL = http://s3.amazonaws.com as a variable for our actions as well, but we need updates on the Github workflow file from the main branch to take this into account. Let's do the following:
> 
>     * Please resolve the merge conflicts
> 
>     * Use the static value `endpoint="http://s3.amazonaws.com"` for these tests
> 
>     * We merge the PR after CI checks pass
> 
>     * I'll take it on our end to update the infrastructure and CI accordingly.
> 
> 
> Is this ok with you?

Sounds good to me. Thank you for the suggestion. I made the change with static endpoint. Once the integration tests pass I will resolve the merge conflicts.

**Comment by dnanuti** [comment]
Thanks for the fast update @bharatendra. Looks like it is failing the Python lints check. Would you please run `black ./` in the content root folder to resolve this?

**Comment by bharatendra** [comment]
> Thanks for the fast update @bharatendra. Looks like it is failing the Python lints check. Would you please run `black ./` in the content root folder to resolve this?

I did merge through the github and introduced those changes. I fixed it from my local branch and checked in. Thank you

**Comment by bharatendra** [comment]
> Thanks for the fast update @bharatendra. Looks like it is failing the Python lints check. Would you please run `black ./` in the content root folder to resolve this?

Looks like the S3Express cases are failing in create session

**Comment by dnanuti** [comment]
Hi @bharatendra! Given that we need a separate configuration for S3 vs. S3 Express which is easier to address on our end, as a team we have decided to take in you changes and follow up with the CI fixes today. Thank you for your contribution!

**Review by dnanuti — APPROVED** [review]

**Review by muddyfish — APPROVED** [review]
